Products/CMFPlone/controlpanel/browser/relations.py

```diff
@@ -0,0 +1,612 @@
+                info[rel.to_id].append(rel.from_id)
+            else:
+                info[rel.from_id].append(rel.to_id)
+
+        for column_1_intid in info:
+            obj = intids.getObject(column_1_intid)
+            use_view_action = obj.portal_type in view_action
+            url = obj.absolute_url() + '/view' if use_view_action else obj.absolute_url()
+            item = {}
+            item['column_1'] = {
+                'title': obj.title,
+                'url': url,
+                'portal_type': obj.portal_type,
+            }
+            item['column_2'] = []
+            for column_2_intid in info[column_1_intid]:
+                obj = intids.getObject(column_2_intid)
+                use_view_action = obj.portal_type in view_action
+                url = obj.absolute_url() + '/view' if use_view_action else obj.absolute_url()
+                item['column_2'].append({
+                    'title': obj.title,
+                    'url': url,
+                    'portal_type': obj.portal_type,
+                    })
+            self.relations.append(item)
+
+        return self.index()
+
+
+def rebuild_relations(context=None, flush_and_rebuild_intids=False):
+    store_relations()
+    purge_relations()
+    if flush_and_rebuild_intids:
+        flush_intids()
+        rebuild_intids()
+    else:
+        cleanup_intids()
+    restore_relations()
+
+
+def get_relations_stats():
+    info = defaultdict(int)
+    relation_catalog = getUtility(ICatalog)
+    for rel in relation_catalog.findRelations():
+        if rel.isBroken():
+            info[rel.from_attribute + ' (broken)'] += 1
+        else:
+            info[rel.from_attribute] += 1
+    return info
+
+
+def get_all_relations():
+    """Get all data from zc.relation catalog.
+    Logs some useful statistics.
+    """
+    results = []
+    info = defaultdict(int)
+
+    relation_catalog = getUtility(ICatalog)
+    for rel in relation_catalog.findRelations():
+        if rel.from_object and rel.to_object:
+            try:
+                results.append({
+                    'from_uuid': rel.from_object.UID(),
+                    'to_uuid': rel.to_object.UID(),
+                    'from_attribute': rel.from_attribute,
+                })
+                info[rel.from_attribute] += 1
+            except AttributeError as ex:
+                logger.info(u'Something went wrong while storing {0}: \n {1}'.format(rel, ex))
+        else:
+            logger.info(u'Dropping relation {} from {} to {}'.format(rel.from_attribute, rel.from_object, rel.to_object))
+    msg = ''
+    for k, v in info.items():
+        msg += u'{}: {}\n'.format(k, v)
+    logger.info(u'\nFound the following relations:\n{}'.format(msg))
+    return results
+
+
+def store_relations(context=None):
+    """Store all relations in a annotation on the portal.
+    """
+    all_relations = get_all_relations()
+    portal = api.portal.get()
+    IAnnotations(portal)[RELATIONS_KEY] = all_relations
```

Comment:
The annotation is purged in restore_relations when rebuilding relations. 
If something fails before that during rebuild_relations is running there will be no commit and the old annotation will not be there. If used manually as a helper-method than the developer doing that should be able take care of it.

Reply:
Question, why to use a persistent annotation at all? Would be a volatile attribute or an annotation to the request be sufficient?
I do not like this, because the probability to have data left attached forever is given. If it does not need to live longer than one request, it should be stored accordingly.